Plutarch/Api/V1/Scripts.hs

```diff
@@ -64,3 +113,230 @@ deriving via
+    Term s (PAsData ctx) ->
+    Term s p
+  ) ->
+  ClosedTerm PValidator
+pwrapValidatorAsData = pwrapValidatorWith punsafeAsData punsafeAsData punsafeAsData
+
+{- |
+  Construct a `PValidator`, assuming the `PData` arguments are
+  of the correct form, wrapping them with `pfromData`.
+
+  Shorthand for:
+  @pwrapValidatorWith punsafeFromData punsafeFromData punsafeFromData@
+-}
+pwrapValidatorFromData ::
+  ( PIsData datum
+  , PIsData redeemer
+  , PIsData context
+  ) =>
+  ( forall s.
+    Term s datum ->
+    Term s redeemer ->
+    Term s context ->
+    Term s p
+  ) ->
+  ClosedTerm PValidator
+pwrapValidatorFromData =
+  pwrapValidatorWith punsafeFromData punsafeFromData punsafeFromData
+
+{- |
+  Construct a `PValidator` using the provided wrappers
+  for each of the arguments.
+-}
+pwrapValidatorWith ::
+  forall datum redeemer ctx p.
+  (forall s. Term s PData -> Term s datum) ->
+  (forall s. Term s PData -> Term s redeemer) ->
+  (forall s. Term s PData -> Term s ctx) ->
+  (forall s. Term s datum -> Term s redeemer -> Term s ctx -> Term s p) ->
+  ClosedTerm PValidator
+pwrapValidatorWith wrapD wrapR wrapCtx validator =
+  plam $ \datum redeemer ctx ->
+    popaque $ validator (wrapD datum) (wrapR redeemer) (wrapCtx ctx)
+
+{- |
+  Construct a `PMintingPolicy`, assuming the `PData` arguments are
+  of the correct form.
+
+  Shorthand for:
+  @pwrapMintingPolicyWith punsafeAsData punsafeAsData@
+-}
+pwrapMintingPolicyAsData ::
+  ( forall s.
+    Term s (PAsData redeemer) ->
+    Term s (PAsData ctx) ->
+    Term s p
+  ) ->
+  ClosedTerm PMintingPolicy
+pwrapMintingPolicyAsData = pwrapMintingPolicyWith punsafeAsData punsafeAsData
+
+{- |
+  Construct a `PMintingPolicy`, assuming the `PData` arguments are
+  of the correct form, wrapping them with `pfromData`.
+
+  Shorthand for:
+  @pwrapMintingPolicyWith punsafeFromData punsafeFromData@
+-}
+pwrapMintingPolicyFromData ::
+  ( PIsData redeemer
+  , PIsData context
+  ) =>
+  ( forall s.
+    Term s redeemer ->
+    Term s context ->
+    Term s p
+  ) ->
+  ClosedTerm PMintingPolicy
+pwrapMintingPolicyFromData =
+  pwrapMintingPolicyWith punsafeFromData punsafeFromData
+
+{- |
+  Construct a `PMintingPolicy` using the provided wrappers
+  for each of the arguments.
+-}
+pwrapMintingPolicyWith ::
+  forall redeemer ctx p.
+  (forall s. Term s PData -> Term s redeemer) ->
+  (forall s. Term s PData -> Term s ctx) ->
+  (forall s. Term s redeemer -> Term s ctx -> Term s p) ->
+  ClosedTerm PMintingPolicy
+pwrapMintingPolicyWith wrapR wrapCtx policy =
+  plam $ \redeemer ctx ->
+    popaque $ policy (wrapR redeemer) (wrapCtx ctx)
+
+{- |
+  Construct a `PStakeValidator`, assuming the `PData` arguments are
+  of the correct form.
+
+  Shorthand for:
+  @pwrapStakeValidatorWith punsafeAsData punsafeAsData@
+-}
+pwrapStakeValidatorAsData ::
+  ( forall s.
+    Term s (PAsData redeemer) ->
+    Term s (PAsData ctx) ->
+    Term s p
+  ) ->
+  ClosedTerm PStakeValidator
```

Comment:
Equivalent to pwrapMintingPolicy... we could avoid the redundancy but it's nice to have separately named versions IMO